net/nimble/host/services/bas/src/ble_svc_bas.c

```diff
@@ -0,0 +1,148 @@
+#if MYNEWT_VAL(BLE_SVC_BAS_BATTERY_LEVEL_NOTIFY) > 0
+static uint16_t ble_svc_bas_battery_handle;
+#endif
+
+/* Connection handle 
+ *
+ * TODO: In order to support multiple connections we would need to save
+ *       the handles for every connection, not just the most recent. Then
+ *       we would need to notify each connection when needed.
+ * */
+static uint16_t ble_svc_bas_conn_handle;
+
+/* Battery level */
+uint8_t ble_svc_bas_battery_level;
+
+/* Access function */
+static int
+ble_svc_bas_access(uint16_t conn_handle, uint16_t attr_handle,
+                   struct ble_gatt_access_ctxt *ctxt, void *arg);
+
+static const struct ble_gatt_svc_def ble_svc_bas_defs[] = {
+    {
+        /*** Battery Service. */
+        .type = BLE_GATT_SVC_TYPE_PRIMARY,
+        .uuid = BLE_UUID16_DECLARE(BLE_SVC_BAS_UUID16),
+        .characteristics = (struct ble_gatt_chr_def[]) { {
+	    /*** Battery level characteristic */
+            .uuid = BLE_UUID16_DECLARE(BLE_SVC_BAS_CHR_UUID16_BATTERY_LEVEL),
+            .access_cb = ble_svc_bas_access,
+#if MYNEWT_VAL(BLE_SVC_BAS_BATTERY_LEVEL_NOTIFY) > 0
+	    .val_handle = &ble_svc_bas_battery_handle,
+#endif
+            .flags = BLE_GATT_CHR_F_READ |
+	             MYNEWT_VAL(BLE_SVC_BAS_BATTERY_LEVEL_PERM) |
+	             MYNEWT_VAL(BLE_SVC_BAS_BATTERY_LEVEL_NOTIFY),
+	    }, {
+            0, /* No more characteristics in this service. */
+        } },
+    },
+
+    {
+        0, /* No more services. */
+    },
+};
+
+/**
+ * BAS access function
+ */
+static int
+ble_svc_bas_access(uint16_t conn_handle, uint16_t attr_handle,
+                          struct ble_gatt_access_ctxt *ctxt,
+                          void *arg)
+{
+    uint16_t uuid16 = ble_uuid_u16(ctxt->chr->uuid);
+    int rc;
+    
+    switch (uuid16) {
+    case BLE_SVC_BAS_CHR_UUID16_BATTERY_LEVEL:
+        assert(ctxt->op == BLE_GATT_ACCESS_OP_READ_CHR);
+        rc = os_mbuf_append(ctxt->om, &ble_svc_bas_battery_level,
+                            sizeof ble_svc_bas_battery_level);
+        return rc == 0 ? 0 : BLE_ATT_ERR_INSUFFICIENT_RES;
+
+    default:
+        assert(0);
+        return BLE_ATT_ERR_UNLIKELY;
+    }
+}
+
+/**
+ * This function must be called with the connection handle when a gap 
+ * connect event is received in order to send notifications to the
+ * client.
+ *
+ * @params conn_handle          The connection handle for the current
+ *                                  connection.
+ */
+void 
+ble_svc_bas_on_gap_connect(uint16_t conn_handle) 
+{
+    ble_svc_bas_conn_handle = conn_handle;
+}
+
+/**
+ * Set the battery level, must be between 0 and 100.
+ * If greater than 100, it will be silently set to 100.
+ */
+int
+ble_svc_bas_battery_level_set(uint8_t level) {
+    if (level > 100)
+	level = 100;
+    ble_svc_bas_battery_level = level;
+#if MYNEWT_VAL(BLE_SVC_BAS_BATTERY_LEVEL_NOTIFY) > 0
+    return ble_gattc_notify(ble_svc_bas_conn_handle, 
+                            ble_svc_bas_battery_handle);
```

Comment:
we probably should not notify if value did not change, i.e. if ble_svc_bas_battery_level == level